### ISSUE TITLE
[Reproduction] AgentOPSD gains are not recovered on fixed ALFWorld/WebShop validation panels

### ISSUE THREAD

**Issue by zhangdw156** [ZethWang/AgentOPSD#11]
Thank you for releasing AgentOPSD. I am trying to reproduce the paper and would appreciate help identifying the remaining implementation or protocol difference.

## Provenance

I created [`zhangdw156/AgentOPSD`](https://github.com/zhangdw156/AgentOPSD/tree/exp/iclr) from official `master` commit [`0c478b2`](https://github.com/ZethWang/AgentOPSD/commit/0c478b2d7cdc201d9b1f076ec5b3dec7e88a161b). The baseline snapshot has the same Git tree as that official commit, and the AgentOPSD math in `opsd_utils.py`, the episode reward manager, and both benchmark prompts remain byte-identical at my experiment HEAD [`2022d1a`](https://github.com/zhangdw156/AgentOPSD/commit/2022d1ad34215e24d53299e76c4b78930a4af12a). Because the released trainer file was incomplete, as also reported in [#10](https://github.com/ZethWang/AgentOPSD/issues/10), I reconstructed/refactored the trainer hook path while preserving the static call order: teacher log-probabilities, reference/reward, GRPO advantage, then AgentOPSD reshaping.

My other changes add environment fairness/evaluation, 1.5B and four-GPU launchers, active-rollout compaction, WebShop lifecycle fixes, an engine-only vLLM seed fix, and logging cleanup. The initial fairness snapshot is [`ef39fea`](https://github.com/zhangdw156/AgentOPSD/commit/ef39fea0db972e13782fdfb23e6e844219acf199).

## Why I added a fairness layer

The inherited environment path did not define a machine-independent task panel. ALFWorld filesystem discovery and independent worker-local validation streams could change task membership/order and repeat tasks; WebShop worker seeds could change the full content represented by a numeric `goal_idx`; and resumed runs restarted worker-local validation streams. I identified these behaviors through a local source audit and cross-machine checks, then tracked the fixes in the method fork history.

The fairness layer uses explicit manifests and stable identities: 3,553 ALFWorld training games, 140 `val_seen`, 134 `val_unseen`, 6,410 WebShop training goals, and 500 WebShop validation goals. It validates uniqueness/counts, uses stable relative ALFWorld paths, makes WebShop goal/price construction worker-independent and fingerprint-checked, samples WebShop training tasks without replacement before group replication, and evaluates each fixed validation task exactly once in deterministic chunks. The 16-task × 8-rollout training contract is unchanged.

This is intentionally a controlled comparison protocol rather than a claim of exact trajectory parity with the paper. There is no fairness-on/off factorial yet, so the curves do not establish that fairness caused the gap.

## Experiment protocol

- Qwen2.5-Instruct at 1.5B, 3B, and 7B; one training seed per cell; 150 updates.
- Task batch 16, group size 8, learning rate `1e-6`, one PPO epoch, validation every five updates.
- ALFWorld: fixed 140-task `val_seen` panel; plotted metric is the unweighted macro-average of six family success rates.
- WebShop: fixed 500-task panel; plotted metric is exact success.
- One stochastic validation sample per task at temperature 0.4.
- GRPO uses the same task pools/evaluator but was trained from a separate baseline implementation, so byte-identical base framework code is not established.

## Results

The paper reports AgentOPSD/GRPO ALFWorld success of `84.4/75.0` at 3B and `89.1/81.2` at 7B, and WebShop exact success of `69.5/63.3` and `79.7/72.6`. My step-150 endpoints are:

| Scale | ALFWorld AgentOPSD / GRPO | WebShop AgentOPSD / GRPO |
|---|---:|---:|
| 1.5B | 68.6 / 61.9 | 58.4 / 62.8 |
| 3B | 62.6 / 65.9 | 63.2 / 63.6 |
| 7B | 78.9 / 80.3 | 69.2 / 73.8 |

The 1.5B row is an extension. AgentOPSD reaches a much higher ALFWorld 3B peak (`74.6` at update 140) than its endpoint, but its normalized validation AUC is still below GRPO (`46.5` versus `48.4`).

## Details I may be interpreting incorrectly

1. Paper Eq. (8) initializes the empirical prior with binary success rate `S/G`, but the released helper receives raw sparse rewards. ALFWorld/WebShop success is encoded as 10, so averaging raw rewards and clamping to `[1e-4, 0.9999]` makes any mixed group with at least one success saturate near one rather than starting at `S/8`.
2. The paper states belief decay `gamma=0.95`; the released launcher does not set it and the helper/trainer default is `1.0`.
3. The paper describes turn-level AgentOPSD for both tasks; the released WebShop launcher sets token granularity.
4. The paper gives PPO clipping `0.2/0.24`; the released/effective setting is `0.2/0.2`.
5. Episode rewards are duplicated onto turn rows and the released GRPO path normalizes over those rows by default, which weights longer trajectories more heavily than a one-advantage-per-trajectory reading.
6. The released default skill placement prepends the skill before the ChatML system message, whereas the paper prompt figure places it in the user task prompt.

## Questions

1. Could you publish the complete runnable trainer commit used for Table 1, including the missing trainer code discussed in #10?
2. Is the intended prior exactly binary `S/G`? If so, should repository rewards be divided by 10 before `compute_group_mean_v0`, and were the paper runs trained that way?
3. Should faithful reproduction use `gamma=0.95`, turn-level WebShop credit, and upper clip 0.24, or the public launcher defaults?
4. Could you share the exact ALFWorld/WebShop task IDs, model/tokenizer revisions, environment versions, seeds, and final-versus-best checkpoint rule used for the reported numbers?
5. Does the recursive evidence sum require normalization by turn/token count when response lengths differ across model scales?
6. Are there known interactions with canonical fixed-task scheduling or exhaustive validation that would explain the smaller gains?

I can provide resolved configs, full scalar CSVs, logs, and a source-level diff audit. Any correction to my reading or a minimal known-good command would be very helpful.

**Comment by zhangdw156** [comment]
<img width="2517" height="2357" alt="Image" src="https://github.com/user-attachments/assets/7b025b13-6f0c-46d3-addf-02f26b2c83dc" />

**Comment by yysirs** [comment]
I was wondering, have you had any successful reproductions yet?

**Comment by zhangdw156** [comment]
> I was wondering, have you had any successful reproductions yet?

None. I'm very confused.
